test/e2e/auth/audit_dynamic.go

```diff
@@ -0,0 +1,381 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	auditv1 "k8s.io/apiserver/pkg/apis/audit/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/utils"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = SIGDescribe("[Feature:DynamicAudit]", func() {
+	f := framework.NewDefaultFramework("audit")
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce")
+	})
+
+	// TODO: Get rid of [DisabledForLargeClusters] when feature request #53455 is ready.
+	It("should dynamically audit API calls [DisabledForLargeClusters]", func() {
+		namespace := f.Namespace.Name
+
+		By("Creating a kubernetes client that impersonates an unauthorized anonymous user")
+		config, err := framework.LoadConfig()
+		framework.ExpectNoError(err, "failed to fetch config")
+
+		config.Impersonate = restclient.ImpersonationConfig{
+			UserName: "system:anonymous",
+			Groups:   []string{"system:unauthenticated"},
+		}
+		anonymousClient, err := clientset.NewForConfig(config)
+		framework.ExpectNoError(err, "failed to create the anonymous client")
+
+		_, err = f.ClientSet.CoreV1().Namespaces().Create(&apiv1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "audit",
+			},
+		})
+		framework.ExpectNoError(err, "failed to create namespace")
+
+		_, err = f.ClientSet.CoreV1().Pods(namespace).Create(&apiv1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "audit-proxy",
+				Labels: map[string]string{
+					"app": "audit",
+				},
+			},
+			Spec: apiv1.PodSpec{
+				Containers: []apiv1.Container{
+					{
+						Name:  "proxy",
+						Image: imageutils.GetE2EImage(imageutils.AuditProxy),
+						Ports: []apiv1.ContainerPort{
+							{
+								ContainerPort: 8080,
+							},
+						},
+					},
+				},
+			},
+		})
+		framework.ExpectNoError(err, "failed to create proxy pod")
+
+		_, err = f.ClientSet.CoreV1().Services(namespace).Create(&apiv1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "audit",
+			},
+			Spec: apiv1.ServiceSpec{
+				Ports: []apiv1.ServicePort{
+					{
+						Port:       80,
+						TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: 8080},
+					},
+				},
+				Selector: map[string]string{
+					"app": "audit",
+				},
+			},
+		})
+		framework.ExpectNoError(err, "failed to create proxy service")
+
+		config, err = framework.LoadConfig()
+		framework.ExpectNoError(err, "failed to load config")
+
+		var podIP string
+		// get pod ip
+		err = wait.Poll(100*time.Millisecond, 10*time.Second, func() (done bool, err error) {
+			p, err := f.ClientSet.CoreV1().Pods(namespace).Get("audit-proxy", metav1.GetOptions{})
+			if errors.IsNotFound(err) {
+				return false, nil
+			} else if err != nil {
+				return false, err
+			}
+			podIP = p.Status.PodIP
+			if podIP == "" {
+				return false, nil
```

Comment:
Don't we need to output logs which explain the IP is not ready yet for digging the issue when the timeout happens for distinguishing a pod nonexistence issue(line 119) or the IP issue?

Reply:
that sounds like a good idea, adding

Reply:
Thanks for updating :)

Reply:
Thanks for updating :-)